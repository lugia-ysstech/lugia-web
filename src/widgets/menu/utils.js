/**
 * create by szfeng
 *
 * @flow
 */
import type { MenuProps, MenuState } from './index';
import * as React from 'react';

const EmptyData = [];

export function isRoot(level: number): boolean {
  return level === 0;
}

export function getSelectedKeys(props: MenuProps, state: ?MenuState): Array<string> {
  const { selectedKeys = [], defaultSelectedKeys = [] } = props;
  if ('selectedKeys' in props) {
    return selectedKeys;
  } else if ('defaultSelectedKeys' in props) {
    return defaultSelectedKeys;
  }
  return state ? state.selectedKeys : [];
}

export function getPopupVisible(props: MenuProps, state: MenuState): boolean {
  const { popupVisible = true } = props;
  if ('popupVisible' in props) {
    return popupVisible;
  }

  return true;
}

export function getTargetOrDefaultTarget(condition: boolean, target: any, defaultTarget: any) {
  return condition ? target : defaultTarget;
}

export function getTargetOrDefaultTargetLazy(
  condition: boolean,
  target: Function,
  defaultTarget: Function
) {
  const matchFunc = getTargetOrDefaultTarget(condition, target, defaultTarget);
  return matchFunc();
}

export function getexpandedPathInProps(props: MenuProps) {
  return 'expandedPath' in props;
}

export function getExpandedPath(props: MenuProps, state: ?MenuState): Array<string> {
  const { expandedPath = [] } = props;
  if (getexpandedPathInProps(props)) {
    return expandedPath;
  }
  return state ? state.expandedPath : [];
}

export function recurDataAndGetSelectedKeys(
  inData: Object[],
  inSelectedKeys: string[],
  outSelectedKeys: string[] = []
): void {
  if (!inSelectedKeys || inSelectedKeys.length === 0) {
    return;
  }

  const target = inSelectedKeys[0];
  inData &&
    inData.forEach(item => {
      const { value } = item;
      if (value === target) {
        outSelectedKeys.push(value);
        const { children } = item;
        if (children && children.length > 0) {
          inSelectedKeys.splice(0, 1);
          recurDataAndGetSelectedKeys(children, inSelectedKeys, outSelectedKeys);
        }
      }
    });
}

export function getCascaderKeys(targetArray: string[] = [], separator: string): string[] {
  if (!targetArray || targetArray.length === 0) {
    return [];
  }

  return targetArray[0].split(separator);
}

export const getExpandDataOrSelectData = letExpandpathOrSelectedKeysToArray;

export function letExpandpathOrSelectedKeysToArray(
  props: MenuProps,
  target: string[] = []
): string[] {
  const result = [];

  if (!props || !target || target.length === 0) {
    return result;
  }

  const { separator, data } = props;
  const cascaderKeys = getCascaderKeys(target, separator);
  recurDataAndGetSelectedKeys(data, cascaderKeys, result);
  return result;
}

export function getExpandedPathFromPropsOrState(props: MenuProps, state: MenuState) {
  const isExpandedPath = 'expandedPath' in props;
  const { expandedPath: stateExpandedPath = [] } = state;
  const { expandedPath: propsExpandedPath = [] } = props;

  return isExpandedPath ? propsExpandedPath : stateExpandedPath;
}

export function getExpandedData(props: MenuProps, state: MenuState) {
  const { expandedData = [], level } = props;

  const rootExpandedPath = getExpandedPathFromPropsOrState(props, state);
  return getTargetOrDefaultTargetLazy(
    isRoot(level),
    () => getExpandDataOrSelectData(props, rootExpandedPath),
    () => expandedData
  );
}

export function getChildData(props: MenuProps, state: MenuState) {
  const { level, allChildData } = props;
  const rootExpandedPath = getExpandedData(props, state);

  if (rootExpandedPath.length === 0) {
    return EmptyData;
  }
  const activeAllChildData = getTargetOrDefaultTargetLazy(
    isRoot(level),
    () => getInitAllChildData(props, state),
    () => allChildData
  );
  return activeAllChildData[level];
}

export function getInitChildData(props: MenuProps, state: MenuState | null): any {
  if (!state) {
    return EmptyData;
  }
  if (state) {
    const rootExpandedPath = getExpandedPathFromPropsOrState(props, state);

    if (rootExpandedPath.length === 0) {
      return EmptyData;
    }
    return getChildData(props, state) || EmptyData;
  }
}

/**
 * 获取所有的childrenData
 */

export function getInitAllChildData(props: MenuProps, state: MenuState): Object {
  const { data } = props;
  return mapGetAllChildData(data, getExpandedData(props, state), 0);
}

export function mapGetAllChildData(
  levelData: Object[],
  expandedData: string[],
  level: number
): Object {
  let target = {};
  if (!expandedData || expandedData.length === 0) {
    return target;
  }
  const expandKey = expandedData[level];

  const expandItem =
    levelData &&
    levelData.find(item => {
      const { value, children } = item;
      return value === expandKey && children && children.length > 0;
    });

  if (expandItem) {
    const { children } = expandItem;
    target[level] = children;
    target = { ...target, ...mapGetAllChildData(children, expandedData, level + 1) };
  }
  return target;
}

export function getTreeData(props: Object, pathSeparator: string = '/') {
  const newData = [];
  const {
    data,
    valueField = 'value',
    displayField = 'text',
    pathField = 'path',
    pidField = 'pid',
  } = props;
  if (data && data.length > 0) {
    recurTreeData(
      data,
      newData,
      {},
      { displayField, valueField, pathSeparator, pathField, pidField }
    );
  }
  return newData;
}

export function recurTreeData(
  inTreeChildData: Object[],
  outTreeRowData: Object[],
  recursion: {
    parentKey?: string,
    parentPath?: string[],
  },

  opt?: {
    onAdd?: Function,
    displayField?: string,
    valueField?: string,
    pathSeparator: string,
    pathField: string,
    pidField: string,
  } = {}
) {
  const { parentKey, parentPath = [] } = recursion;
  const {
    displayField = 'text',
    valueField = 'value',
    pathSeparator,
    pathField = 'path',
    pidField = 'pid',
  } = opt;
  const { onAdd } = opt;

  inTreeChildData &&
    inTreeChildData.forEach(item => {
      const { children, [valueField]: value, [displayField]: text, ...other } = item;
      const newObj = { ...other };
      newObj[pidField] = parentKey;
      newObj[valueField] = value + '';
      newObj[displayField] = text;

      let path;
      if (!parentKey) {
        path = [];
      } else {
        if (parentPath.indexOf(parentKey) === -1) {
          parentPath.push(parentKey);
        }
        path = [...parentPath];
        newObj[pathField] = path.join(pathSeparator);
      }

      const isLeaf = (newObj.isLeaf = !children || children.length === 0);
      outTreeRowData.push(newObj);
      onAdd && onAdd(newObj);
      if (!isLeaf) {
        recurTreeData(
          children,
          outTreeRowData,
          { parentKey: item[valueField] + '', parentPath: path },
          opt
        );
      }
    });
}
