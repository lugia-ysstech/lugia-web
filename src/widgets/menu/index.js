import * as React from 'react';
import Menu from './menu';
import { getMenuItemHeight } from '../css/menu';
import ThemeHoc from '@lugia/theme-hoc';
import Widget from '../consts/index';
import Item from './item';
import { getTreeData } from './utils';

class ThrottleMenu extends React.Component<any, any> {
  constructor(props: any) {
    super(props);
    this.innerMenu = React.createRef();
  }

  render() {
    const { props } = this;
    const { children, getPartOfThemeConfig, __treeData__, separator = '|' } = props;
    const treeData =
      __treeData__ && __treeData__.length !== 0 ? __treeData__ : getTreeData(props, separator);
    const menuItemThemeConfig = getPartOfThemeConfig('MenuItem');
    const menuItemHeight = getMenuItemHeight(menuItemThemeConfig, props);
    return (
      <Menu ref={this.innerMenu} {...props} menuItemHeight={menuItemHeight} treeData={treeData}>
        {children}
      </Menu>
    );
  }
}

const MenuWrap = ThemeHoc(ThrottleMenu, Widget.Menu, {
  hover: true,
});

MenuWrap.MenuItem = Item;
export default MenuWrap;
