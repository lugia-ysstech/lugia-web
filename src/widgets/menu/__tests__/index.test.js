//@flow

import * as React from 'react';
import chai from 'chai';
import Menu from '../index.js';
import Widget from '../../consts/index';
import 'jest-styled-components';
import renderer from 'react-test-renderer';
import Theme from '../../theme';

import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { ItemWrap } from '../../css/menu';

Enzyme.configure({ adapter: new Adapter() });

const { MenuItem } = Menu;
const { expect: exp } = chai;

describe('Menu', () => {
  it('DropMenu single selectedKeys 2', () => {
    expect(
      renderer
        .create(
          <Menu selectedKeys={['2', '3']}>
            <MenuItem key="1">a</MenuItem>
            <MenuItem key="2">b</MenuItem>
            <MenuItem key="3">c</MenuItem>
            <MenuItem key="4">d</MenuItem>
          </Menu>
        )
        .toJSON()
    ).toMatchSnapshot();
  });
  it('DropMenu width: 500px height: 300px', () => {
    expect(
      renderer
        .create(
          <Theme
            config={{
              [Widget.Menu]: {
                Container: {
                  width: 500,
                  height: 300,
                },
              },
            }}
          >
            <Menu>
              <MenuItem key="1">a</MenuItem>
              <MenuItem key="2">b</MenuItem>
              <MenuItem key="3">c</MenuItem>
              <MenuItem key="4">d</MenuItem>
            </Menu>
          </Theme>
        )
        .toJSON()
    ).toMatchSnapshot();
  });

  it('DropMenu mutliple selectedKeys 2', () => {
    expect(
      renderer
        .create(
          <Menu selectedKeys={['3', '4']} mutliple>
            <MenuItem key="1">a</MenuItem>
            <MenuItem key="2">b</MenuItem>
            <MenuItem key="3">c</MenuItem>
            <MenuItem key="4">d</MenuItem>
          </Menu>
        )
        .toJSON()
    ).toMatchSnapshot();
  });

  it('DropMenu mutliple onClick', () => {
    const checkedKey = '4';
    const dom = mount(
      <Menu mutliple>
        <MenuItem key="1">a</MenuItem>
        <MenuItem key="2">b</MenuItem>
        <MenuItem key="3">c</MenuItem>
        <MenuItem key={checkedKey} checked></MenuItem>
      </Menu>
    );

    dom.find('ItemWrap').forEach(node => {
      exp(node.prop('checked')).to.be.false;
    });
    dom.find('ItemWrap').forEach(node => node.simulate('click'));
    dom.find('ItemWrap').forEach(node => {
      exp(node.prop('checked')).to.be.true;
    });
  });
  it('DropMenu mutliple onClick selectedKeys: 1 2', () => {
    const checkedKey = '4';
    const dom = mount(
      <Menu mutliple selectedKeys={['1', '2']}>
        <MenuItem key="1">a</MenuItem>
        <MenuItem key="2">b</MenuItem>
        <MenuItem key="3">c</MenuItem>
        <MenuItem key={checkedKey} checked>
          d
        </MenuItem>
      </Menu>
    );

    exp(
      dom
        .find('ItemWrap')
        .at(0)
        .prop('checked')
    ).to.be.true;
    exp(
      dom
        .find('ItemWrap')
        .at(1)
        .prop('checked')
    ).to.be.true;
    exp(
      dom
        .find('ItemWrap')
        .at(2)
        .prop('checked')
    ).to.be.false;
    exp(
      dom
        .find('ItemWrap')
        .at(3)
        .prop('checked')
    ).to.be.false;
    dom.find('ItemWrap').forEach(node => node.simulate('click'));
    exp(
      dom
        .find('ItemWrap')
        .at(0)
        .prop('checked')
    ).to.be.true;
    exp(
      dom
        .find('ItemWrap')
        .at(1)
        .prop('checked')
    ).to.be.true;
    exp(
      dom
        .find('ItemWrap')
        .at(2)
        .prop('checked')
    ).to.be.false;
    exp(
      dom
        .find('ItemWrap')
        .at(3)
        .prop('checked')
    ).to.be.false;
  });

  it('DropMenu single onClick selectedKeys: 1 2', () => {
    const checkedKey = '4';
    const dom = mount(
      <Menu selectedKeys={['1', '2']}>
        <MenuItem key="1">a</MenuItem>
        <MenuItem key="2">b</MenuItem>
        <MenuItem key="3">c</MenuItem>
        <MenuItem key={checkedKey} checked>
          d
        </MenuItem>
      </Menu>
    );
    exp(
      dom
        .find('ItemWrap')
        .at(0)
        .prop('checked')
    ).to.be.false;
    exp(
      dom
        .find('ItemWrap')
        .at(1)
        .prop('checked')
    ).to.be.true;
    exp(
      dom
        .find('ItemWrap')
        .at(2)
        .prop('checked')
    ).to.be.false;
    exp(
      dom
        .find('ItemWrap')
        .at(3)
        .prop('checked')
    ).to.be.false;
    dom.find('ItemWrap').forEach(node => node.simulate('click'));
    exp(
      dom
        .find('ItemWrap')
        .at(0)
        .prop('checked')
    ).to.be.false;
    exp(
      dom
        .find('ItemWrap')
        .at(1)
        .prop('checked')
    ).to.be.true;
    exp(
      dom
        .find('ItemWrap')
        .at(2)
        .prop('checked')
    ).to.be.false;
    exp(
      dom
        .find('ItemWrap')
        .at(3)
        .prop('checked')
    ).to.be.false;
  });

  it('compute see count', () => {
    const dom = mount(
      <Menu>
        <MenuItem key="1">1</MenuItem>
        <MenuItem key="2">2</MenuItem>
        <MenuItem key="3">3</MenuItem>
        <MenuItem key="4" checked>
          4
        </MenuItem>
        <MenuItem key="5" checked>
          5
        </MenuItem>
        <MenuItem key="6" checked>
          6
        </MenuItem>
        <MenuItem key="7" checked>
          7
        </MenuItem>
        <MenuItem key="8" checked>
          8
        </MenuItem>
        <MenuItem key="9" checked>
          9
        </MenuItem>
        <MenuItem key="11" checked>
          10
        </MenuItem>
        <MenuItem key="12" checked>
          11
        </MenuItem>
        <MenuItem key="13" checked>
          12
        </MenuItem>
        <MenuItem key="13" checked>
          13
        </MenuItem>
      </Menu>
    );

    exp(dom.find('li').length).to.be.equal(8);
  });

  it('children: items', () => {
    const items = [];
    for (let i = 0; i < 10; i++) {
      items.push(<MenuItem key={i}>{i}</MenuItem>);
    }
    const wraper = renderer.create(
      <Theme
        config={{
          [Widget.Menu]: {
            Container: {
              normal: {
                width: 200,
                height: 250,
              },
            },
          },
        }}
      >
        <Menu mutliple>{items}</Menu>
      </Theme>
    );
    expect(wraper.toJSON()).toMatchSnapshot();
  });
});
