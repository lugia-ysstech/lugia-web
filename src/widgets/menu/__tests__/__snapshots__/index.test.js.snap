// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu DropMenu mutliple selectedKeys 2 1`] = `
.c1 {
  outline: none;
  margin: 0;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  padding-left: 0;
  list-style: none;
  overflow: hidden;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  width: 25rem;
  background-color: #fff;
  height: 25rem;
}

.c2 {
  box-sizing: border-box;
  position: relative;
  font-weight: 100;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  -webkit-transition-property: background-color,border,borderRadius,opacity,boxShadow;
  transition-property: background-color,border,borderRadius,opacity,boxShadow;
  -webkit-transition-duration: 0.3s;
  transition-duration: 0.3s;
  font-size: 1.4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  cursor: pointer;
  height: 3.5rem;
}

.c2 > i {
  vertical-align: middle;
}

.c3 {
  padding: 0rem 0.8rem;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  overflow: hidden;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c3 i {
  vertical-align: middle;
}

.c4 {
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
}

.c0 {
  position: relative;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  overflow: hidden;
  border-radius: 0.4rem;
  width: 25rem;
  background-color: #fff;
  height: 25rem;
}

.c0:hover > div:nth-child(2) {
  opacity: 1;
}

<div
  className="c0"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onWheel={[Function]}
  style={
    Object {
      "boxShadow": "0px 1px 6px 0px rgba(0, 0, 0, 0.2)",
    }
  }
>
  <ul
    className="c1"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={Object {}}
  >
    <li
      className="c2"
      disabled={false}
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      style={Object {}}
      title="a"
    >
      <div
        className="c3"
        style={Object {}}
      >
        <span
          className="c4"
        >
          a
        </span>
      </div>
    </li>
    <li
      className="c2"
      disabled={false}
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      style={Object {}}
      title="b"
    >
      <div
        className="c3"
        style={Object {}}
      >
        <span
          className="c4"
        >
          b
        </span>
      </div>
    </li>
    <li
      className="c2"
      disabled={false}
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      style={
        Object {
          "color": "#4d63ff",
          "fontWeight": 900,
        }
      }
      title="c"
    >
      <div
        className="c3"
        style={Object {}}
      >
        <span
          className="c4"
        >
          c
        </span>
      </div>
    </li>
    <li
      className="c2"
      disabled={false}
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      style={
        Object {
          "color": "#4d63ff",
          "fontWeight": 900,
        }
      }
      title="d"
    >
      <div
        className="c3"
        style={Object {}}
      >
        <span
          className="c4"
        >
          d
        </span>
      </div>
    </li>
  </ul>
</div>
`;

exports[`Menu DropMenu single selectedKeys 2 1`] = `
.c1 {
  outline: none;
  margin: 0;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  padding-left: 0;
  list-style: none;
  overflow: hidden;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  width: 25rem;
  background-color: #fff;
  height: 25rem;
}

.c2 {
  box-sizing: border-box;
  position: relative;
  font-weight: 100;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  -webkit-transition-property: background-color,border,borderRadius,opacity,boxShadow;
  transition-property: background-color,border,borderRadius,opacity,boxShadow;
  -webkit-transition-duration: 0.3s;
  transition-duration: 0.3s;
  font-size: 1.4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  cursor: pointer;
  height: 3.5rem;
}

.c2 > i {
  vertical-align: middle;
}

.c3 {
  padding: 0rem 0.8rem;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  overflow: hidden;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c3 i {
  vertical-align: middle;
}

.c4 {
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
}

.c0 {
  position: relative;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  overflow: hidden;
  border-radius: 0.4rem;
  width: 25rem;
  background-color: #fff;
  height: 25rem;
}

.c0:hover > div:nth-child(2) {
  opacity: 1;
}

<div
  className="c0"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onWheel={[Function]}
  style={
    Object {
      "boxShadow": "0px 1px 6px 0px rgba(0, 0, 0, 0.2)",
    }
  }
>
  <ul
    className="c1"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={Object {}}
  >
    <li
      className="c2"
      disabled={false}
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      style={Object {}}
      title="a"
    >
      <div
        className="c3"
        style={Object {}}
      >
        <span
          className="c4"
        >
          a
        </span>
      </div>
    </li>
    <li
      className="c2"
      disabled={false}
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      style={Object {}}
      title="b"
    >
      <div
        className="c3"
        style={Object {}}
      >
        <span
          className="c4"
        >
          b
        </span>
      </div>
    </li>
    <li
      className="c2"
      disabled={false}
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      style={
        Object {
          "color": "#4d63ff",
          "fontWeight": 900,
        }
      }
      title="c"
    >
      <div
        className="c3"
        style={Object {}}
      >
        <span
          className="c4"
        >
          c
        </span>
      </div>
    </li>
    <li
      className="c2"
      disabled={false}
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      style={Object {}}
      title="d"
    >
      <div
        className="c3"
        style={Object {}}
      >
        <span
          className="c4"
        >
          d
        </span>
      </div>
    </li>
  </ul>
</div>
`;

exports[`Menu DropMenu width: 500px height: 333px 1`] = `
.c1 {
  outline: none;
  margin: 0;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  padding-left: 0;
  list-style: none;
  overflow: hidden;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  width: 25rem;
  background-color: #fff;
  height: 25rem;
}

.c2 {
  box-sizing: border-box;
  position: relative;
  font-weight: 100;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  -webkit-transition-property: background-color,border,borderRadius,opacity,boxShadow;
  transition-property: background-color,border,borderRadius,opacity,boxShadow;
  -webkit-transition-duration: 0.3s;
  transition-duration: 0.3s;
  font-size: 1.4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  cursor: pointer;
  height: 3.5rem;
}

.c2 > i {
  vertical-align: middle;
}

.c3 {
  padding: 0rem 0.8rem;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  overflow: hidden;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c3 i {
  vertical-align: middle;
}

.c4 {
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
}

.c0 {
  position: relative;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  overflow: hidden;
  border-radius: 0.4rem;
  width: 25rem;
  background-color: #fff;
  height: 25rem;
}

.c0:hover > div:nth-child(2) {
  opacity: 1;
}

<div
  className="c0"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onWheel={[Function]}
  style={
    Object {
      "boxShadow": "0px 1px 6px 0px rgba(0, 0, 0, 0.2)",
    }
  }
>
  <ul
    className="c1"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={Object {}}
  >
    <li
      className="c2"
      disabled={false}
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      style={Object {}}
      title="a"
    >
      <div
        className="c3"
        style={Object {}}
      >
        <span
          className="c4"
        >
          a
        </span>
      </div>
    </li>
    <li
      className="c2"
      disabled={false}
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      style={Object {}}
      title="b"
    >
      <div
        className="c3"
        style={Object {}}
      >
        <span
          className="c4"
        >
          b
        </span>
      </div>
    </li>
    <li
      className="c2"
      disabled={false}
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      style={Object {}}
      title="c"
    >
      <div
        className="c3"
        style={Object {}}
      >
        <span
          className="c4"
        >
          c
        </span>
      </div>
    </li>
    <li
      className="c2"
      disabled={false}
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      style={Object {}}
      title="d"
    >
      <div
        className="c3"
        style={Object {}}
      >
        <span
          className="c4"
        >
          d
        </span>
      </div>
    </li>
  </ul>
</div>
`;

exports[`Menu children: items 1`] = `
.c2 {
  outline: none;
  margin: 0;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  padding-left: 0;
  list-style: none;
  overflow: hidden;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  width: 25rem;
  background-color: #fff;
  height: 25rem;
}

.c3 {
  box-sizing: border-box;
  position: relative;
  font-weight: 100;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  -webkit-transition-property: background-color,border,borderRadius,opacity,boxShadow;
  transition-property: background-color,border,borderRadius,opacity,boxShadow;
  -webkit-transition-duration: 0.3s;
  transition-duration: 0.3s;
  font-size: 1.4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  cursor: pointer;
  height: 3.5rem;
}

.c3 > i {
  vertical-align: middle;
}

.c4 {
  padding: 0rem 0.8rem;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  overflow: hidden;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c4 i {
  vertical-align: middle;
}

.c5 {
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
}

.c0 {
  position: relative;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  overflow: hidden;
  border-radius: 0.4rem;
  width: 25rem;
  background-color: #fff;
  height: 25rem;
}

.c0:hover > div:nth-child(2) {
  opacity: 1;
}

.c1 {
  width: 25rem;
  position: absolute;
  display: inline-block;
  height: 25rem;
  height: 25rem;
}

.c6 {
  width: 1.2rem;
  position: absolute;
  display: inline-block;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
  opacity: 0;
  left: 23.8rem;
}

.c8 {
  position: relative;
  background: rgba(255,255,255,0);
  width: 2rem;
  height: 30rem;
  z-index: 996;
}

.c7 {
  width: 1.2rem;
}

.c9 {
  position: absolute;
  top: 0;
  left: 0;
  cursor: pointer;
  text-align: center;
  border-radius: 0.8rem;
  background: #c2c2c2;
  color: #fff;
  font-size: 1.2rem;
  line-height: 3rem;
  -webkit-transition: all 0.4s;
  transition: all 0.4s;
  width: 0.8rem;
  margin-left: 0.2rem;
  margin-right: 0.2rem;
}

.c9:hover {
  background-color: #7a7a7a;
}

<div
  className="c0"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onWheel={[Function]}
  style={
    Object {
      "boxShadow": "0px 1px 6px 0px rgba(0, 0, 0, 0.2)",
    }
  }
>
  <div
    className="c1"
    style={Object {}}
  >
    <ul
      className="c2"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      style={Object {}}
    >
      <li
        className="c3"
        disabled={false}
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        style={Object {}}
        title=""
      >
        <div
          className="c4"
          style={Object {}}
        >
          <span
            className="c5"
          >
            0
          </span>
        </div>
      </li>
      <li
        className="c3"
        disabled={false}
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        style={Object {}}
        title=""
      >
        <div
          className="c4"
          style={Object {}}
        >
          <span
            className="c5"
          >
            1
          </span>
        </div>
      </li>
      <li
        className="c3"
        disabled={false}
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        style={Object {}}
        title=""
      >
        <div
          className="c4"
          style={Object {}}
        >
          <span
            className="c5"
          >
            2
          </span>
        </div>
      </li>
      <li
        className="c3"
        disabled={false}
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        style={Object {}}
        title=""
      >
        <div
          className="c4"
          style={Object {}}
        >
          <span
            className="c5"
          >
            3
          </span>
        </div>
      </li>
      <li
        className="c3"
        disabled={false}
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        style={Object {}}
        title=""
      >
        <div
          className="c4"
          style={Object {}}
        >
          <span
            className="c5"
          >
            4
          </span>
        </div>
      </li>
      <li
        className="c3"
        disabled={false}
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        style={Object {}}
        title=""
      >
        <div
          className="c4"
          style={Object {}}
        >
          <span
            className="c5"
          >
            5
          </span>
        </div>
      </li>
      <li
        className="c3"
        disabled={false}
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        style={Object {}}
        title=""
      >
        <div
          className="c4"
          style={Object {}}
        >
          <span
            className="c5"
          >
            6
          </span>
        </div>
      </li>
      <li
        className="c3"
        disabled={false}
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        style={Object {}}
        title=""
      >
        <div
          className="c4"
          style={Object {}}
        >
          <span
            className="c5"
          >
            7
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div
    className="c6"
    style={Object {}}
  >
    <div
      className="c7  c8"
      onMouseDown={[Function]}
      onMouseMove={[Function]}
      onMouseOut={[Function]}
      onMouseUp={[Function]}
      onWheel={[Function]}
      style={
        Object {
          "height": "25rem",
        }
      }
    >
      <div
        className="c9"
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        style={
          Object {
            "height": "17.6rem",
            "top": "0rem",
          }
        }
      />
    </div>
  </div>
</div>
`;
